internal/types/action/registrar.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package action
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/boundary/internal/types/resource"
+)
+
+type resourceActionSets struct {
+	individual ActionSet
+	collection ActionSet
+	valid      ActionSet
+}
+
+type byResource struct {
+	mu sync.RWMutex
+	m  map[resource.Type]*resourceActionSets
+}
+
+func (b *byResource) add(r resource.Type, individual ActionSet, collection ActionSet) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.m == nil {
+		b.m = make(map[resource.Type]*resourceActionSets)
+	}
+
+	_, ok := b.m[r]
+	if ok {
+		return fmt.Errorf("%s already registered", r)
+	}
+	b.m[r] = &resourceActionSets{
+		individual: individual,
+		collection: collection,
+		valid:      append(individual, collection...),
+	}
+
+	return nil
+}
+
+func (b *byResource) get(r resource.Type) (*resourceActionSets, error) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	a, ok := b.m[r]
+	if !ok {
+		return nil, fmt.Errorf("resource not found: %s", r)
+	}
+	return a, nil
+}
+
+var byResourceRegistrar = &byResource{}
+
+// RegisterResource registers the valid ActionSets for a resource.
+func RegisterResource(r resource.Type, individual ActionSet, collection ActionSet) error {
+	return byResourceRegistrar.add(r, individual, collection)
+}
+
+// SetForResource returns the valid ActionSet for the given resource.
+// If the provided resource was not registered, and error is returned.
+func SetForResource(r resource.Type) (ActionSet, error) {
```

Comment:
I like the original `action.SetForResource(...)` over
`action.ResourceActionSet(...)`.  With that said, what about `action.RegisteredActionSet(...)`?

Reply:
To address some TODOs and the comment [here](https://github.com/hashicorp/boundary/pull/3825#discussion_r1373200746), I refactored `action.ActionSet` so it is implemented as a set, see 66e44b75f.

In that commit, I introduce a couple additional package functions that operate on ActionSets. For them I opted to use "ActionSet" in the function name, ie:

- `action.ActionSetUnion`
- `action.ActionSetDifference`

These stutter, which I would generally want to avoid, but it is consistent with `action.ActionSet`. I think I would prefer to either:
1. Be consistent, in which case I think function in question here should be `action.ActionSetForResource`
3. Or fix all of the stutters and refactor `action.ActionSet` to `action.Set`. This would allow this function to be `action.SetForResource` which could still be confusing if you see "Set" and think of a setter. However, since this is a package function rather than a method on a struct, and if there is a type called Set, I don't think it should be that confusing.